File: libs/moqtail-rs/src/model/control/announce_cancel.rs
use bytes::{BufMut, Bytes, BytesMut};

use super::constant::AnnounceErrorCode;
use super::control_message::ControlMessageTrait;
use crate::model::common::reason_phrase::ReasonPhrase;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::control::constant::ControlMessageType;
use crate::model::error::ParseError;

#[derive(Debug, Clone, PartialEq)]
pub struct AnnounceCancel {
  pub track_namespace: Tuple,
  pub error_code: AnnounceErrorCode,
  pub reason_phrase: ReasonPhrase,
}
impl AnnounceCancel {
  pub fn new(
    track_namespace: Tuple,
    error_code: AnnounceErrorCode,
    reason_phrase: ReasonPhrase,
  ) -> Self {
    AnnounceCancel {
      track_namespace,
      error_code,
      reason_phrase,
    }
  }
}
impl ControlMessageTrait for AnnounceCancel {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::AnnounceCancel)?;

    let mut payload = BytesMut::new();
    payload.extend_from_slice(&self.track_namespace.serialize()?);
    payload.put_vi(self.error_code)?;
    payload.extend_from_slice(&self.reason_phrase.serialize()?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "AnnounceCancel::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);
    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let track_namespace = Tuple::deserialize(payload)?;

    let error_code_raw = payload.get_vi()?;
    let error_code = AnnounceErrorCode::try_from(error_code_raw)?;
    let reason_phrase = ReasonPhrase::deserialize(payload)?;

    Ok(Box::new(AnnounceCancel {
      track_namespace,
      error_code,
      reason_phrase,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::AnnounceCancel
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;
  #[test]
  fn test_roundtrip() {
    let error_code = AnnounceErrorCode::ExpiredAuthToken;
    let reason_phrase = ReasonPhrase::try_new("why are you running?".to_string()).unwrap();
    let track_namespace = Tuple::from_utf8_path("valid/track/namespace");
    let sub_update = AnnounceCancel {
      error_code,
      reason_phrase,
      track_namespace,
    };
    let mut buf = sub_update.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::AnnounceCancel as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = AnnounceCancel::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, sub_update);
    assert!(!buf.has_remaining());
  }
  #[test]
  fn test_excess_roundtrip() {
    let error_code = AnnounceErrorCode::InternalError;
    let reason_phrase = ReasonPhrase::try_new("bomboclad".to_string()).unwrap();
    let track_namespace = Tuple::from_utf8_path("another/valid/track/namespace");
    let sub_update = AnnounceCancel {
      error_code,
      reason_phrase,
      track_namespace,
    };
    let serialized = sub_update.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::AnnounceCancel as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = AnnounceCancel::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, sub_update);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let error_code = AnnounceErrorCode::MalformedAuthToken;
    let reason_phrase = ReasonPhrase::try_new("Uvuvwevwevwe".to_string()).unwrap();
    let track_namespace = Tuple::from_utf8_path("Onyetenyevwe/Ugwemuhwem/Osas");
    let sub_update = AnnounceCancel {
      error_code,
      reason_phrase,
      track_namespace,
    };
    let mut buf = sub_update.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::AnnounceCancel as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = AnnounceCancel::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}

File: libs/moqtail-ts/src/model/control/announce_cancel.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Tuple } from '../common/tuple'
import { ReasonPhrase } from '../common/reason_phrase'
import { ControlMessageType, AnnounceErrorCode } from './constant'
import { LengthExceedsMaxError } from '../error/error'

export class AnnounceCancel {
  constructor(
    public readonly trackNamespace: Tuple,
    public readonly errorCode: AnnounceErrorCode,
    public readonly reasonPhrase: ReasonPhrase,
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.AnnounceCancel
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.AnnounceCancel)
    const payload = new ByteBuffer()
    payload.putTuple(this.trackNamespace)
    payload.putVI(this.errorCode)
    payload.putReasonPhrase(this.reasonPhrase)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('AnnounceCancel::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): AnnounceCancel {
    const trackNamespace = buf.getTuple()
    const errorCodeRaw = buf.getVI()
    const errorCode = AnnounceErrorCode.tryFrom(errorCodeRaw)
    const reasonPhrase = buf.getReasonPhrase()
    return new AnnounceCancel(trackNamespace, errorCode, reasonPhrase)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('AnnounceCancel', () => {
    test('roundtrip', () => {
      const errorCode = AnnounceErrorCode.ExpiredAuthToken
      const reasonPhrase = new ReasonPhrase('why are you running?')
      const trackNamespace = Tuple.fromUtf8Path('valid/track/namespace')
      const msg = new AnnounceCancel(trackNamespace, errorCode, reasonPhrase)
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.AnnounceCancel))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = AnnounceCancel.parsePayload(frozen)
      expect(deserialized.trackNamespace.equals(msg.trackNamespace)).toBe(true)
      expect(deserialized.errorCode).toBe(msg.errorCode)
      expect(deserialized.reasonPhrase.phrase).toBe(msg.reasonPhrase.phrase)
      expect(frozen.remaining).toBe(0)
    })
    test('excess roundtrip', () => {
      const errorCode = AnnounceErrorCode.InternalError
      const reasonPhrase = new ReasonPhrase('bomboclad')
      const trackNamespace = Tuple.fromUtf8Path('another/valid/track/namespace')
      const msg = new AnnounceCancel(trackNamespace, errorCode, reasonPhrase)
      const serialized = msg.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.AnnounceCancel))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = AnnounceCancel.parsePayload(frozen)
      expect(deserialized.trackNamespace.equals(msg.trackNamespace)).toBe(true)
      expect(deserialized.errorCode).toBe(msg.errorCode)
      expect(deserialized.reasonPhrase.phrase).toBe(msg.reasonPhrase.phrase)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })
    test('partial message', () => {
      const errorCode = AnnounceErrorCode.MalformedAuthToken
      const reasonPhrase = new ReasonPhrase('Uvuvwevwevwe')
      const trackNamespace = Tuple.fromUtf8Path('Onyetenyevwe/Ugwemuhwem/Osas')
      const msg = new AnnounceCancel(trackNamespace, errorCode, reasonPhrase)
      const serialized = msg.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        AnnounceCancel.parsePayload(frozen)
      }).toThrow()
    })
  })
}
